pyomo/solvers/plugins/solvers/mosek_direct.py

```diff
@@ -163,21 +163,27 @@ def _process_stream(msg):
         for key, option in self.options.items():
             try:
                 param = key.split('.')
-                if param[0] == 'mosek':
-                    param.pop(0)
-                param = getattr(mosek, param[0])(param[1])
-                if 'sparam' in key.split('.'):
-                    self._solver_model.putstrparam(param, option)
-                elif 'dparam' in key.split('.'):
-                    self._solver_model.putdouparam(param, option)
-                elif 'iparam' in key.split('.'):
-                    if isinstance(option, str):
-                        option = option.split('.')
-                        if option[0] == 'mosek':
-                            option.pop('mosek')
-                        option = getattr(mosek, option[0])(option[1])
-                    else:
-                        self._solver_model.putintparam(param, option)
+                if key == param[0]:
+                    self._solver_model.putparam(key, option)
+                else:
+                    if param[0] == 'mosek':
+                        param.pop(0)
+                    param = getattr(mosek, param[0])(param[1])
+                    if 'sparam' in key.split('.'):
+                        self._solver_model.putstrparam(param, option)
+                    elif 'dparam' in key.split('.'):
+                        self._solver_model.putdouparam(param, option)
+                    elif 'iparam' in key.split('.'):
+                        if isinstance(option, int):
+                            self._solver_model.putintparam(param, option)
+                        elif isinstance(option, str):
+                            option = option.split('.')
+                            if option[0] == 'mosek':
+                                option.pop(0)
+                            option = getattr(mosek, option[0])(option[1])
+                            self._solver_model.putintparam(param, option)
+                        else:
+                            logger.warning("Ignoring invalid value for iparam.\n")
```

Comment:
Ideally, this message would include the name of the offending param (i.e., `key`), and the (invalid) type (e.g., `type(option).__name__`)

Reply:
Sorry for flip-flopping like this, but your comment made me think that it is best if I simply raise TypeError in case a value with an incorrect type is passed. The Error message thus raised by MOSEK will contain the necessary information as you have suggested. Furthermore, this makes the way iparams are handled similar to the sparam/dpararm.

Reply:
That seems quite reasonable.